### PR TITLE
Add script to detect circular dependencies between source modules

### PR DESCRIPTION
**Pull request by sipa**
This script finds dependencies between source code modules, treating the `.cpp` and `.h` file as one unit (so it will detect `A.cpp` depending on `B.h` where `B.cpp` depends on `A.h`). This can be used to find out which modules cannot be used independently from each other.

It is very simplistic at this point, and assumes that a `.cpp` file's corresponding header has the exact same name, with `.cpp` replaced by `.h`. Furthermore, it assumes all `#include`s are relative to the `src/` directory.

This is not a linter, and is not enforced through Travis or otherwise.

This is the current output:

```
$ ../contrib/devtools/circular-dependencies.py {*,*/*,*/*/*}.{h,cpp}
Circular dependency: chain -> pow -> chain
Circular dependency: chainparamsbase -> util -> chainparamsbase
Circular dependency: checkpoints -> validation -> checkpoints
Circular dependency: init -> index/txindex -> init
Circular dependency: init -> validation -> init
Circular dependency: init -> net_processing -> init
Circular dependency: init -> rpc/server -> init
Circular dependency: init -> txdb -> init
Circular dependency: init -> validationinterface -> init
Circular dependency: random -> util -> random
Circular dependency: sync -> util -> sync
Circular dependency: txmempool -> validation -> txmempool
Circular dependency: txmempool -> policy/fees -> txmempool
Circular dependency: validation -> index/txindex -> validation
Circular dependency: validation -> policy/policy -> validation
Circular dependency: validation -> validationinterface -> validation
Circular dependency: qt/addresstablemodel -> qt/walletmodel -> qt/addresstablemodel
Circular dependency: qt/bantablemodel -> qt/clientmodel -> qt/bantablemodel
Circular dependency: qt/bitcoingui -> qt/walletview -> qt/bitcoingui
Circular dependency: qt/bitcoingui -> qt/walletframe -> qt/bitcoingui
Circular dependency: qt/bitcoingui -> qt/utilitydialog -> qt/bitcoingui
Circular dependency: qt/clientmodel -> qt/peertablemodel -> qt/clientmodel
Circular dependency: qt/paymentserver -> qt/walletmodel -> qt/paymentserver
Circular dependency: qt/recentrequeststablemodel -> qt/walletmodel -> qt/recentrequeststablemodel
Circular dependency: qt/sendcoinsdialog -> qt/walletmodel -> qt/sendcoinsdialog
Circular dependency: qt/transactiontablemodel -> qt/walletmodel -> qt/transactiontablemodel
Circular dependency: qt/walletmodel -> qt/walletmodeltransaction -> qt/walletmodel
Circular dependency: rpc/rawtransaction -> wallet/rpcwallet -> rpc/rawtransaction
Circular dependency: wallet/coincontrol -> wallet/wallet -> wallet/coincontrol
Circular dependency: wallet/fees -> wallet/wallet -> wallet/fees
Circular dependency: wallet/rpcwallet -> wallet/wallet -> wallet/rpcwallet
Circular dependency: wallet/walletdb -> wallet/wallet -> wallet/walletdb
Circular dependency: txmempool -> validation -> policy/rbf -> txmempool
Circular dependency: txmempool -> validation -> validationinterface -> txmempool
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/addressbookpage
Circular dependency: qt/guiutil -> qt/walletmodel -> qt/optionsmodel -> qt/guiutil
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/signverifymessagedialog -> qt/addressbookpage
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/receivecoinsdialog -> qt/addressbookpage
Circular dependency: qt/guiutil -> qt/walletmodel -> qt/optionsmodel -> qt/intro -> qt/guiutil
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/sendcoinsdialog -> qt/sendcoinsentry -> qt/addressbookpage
```

### CODE REVIEW

**Comment by promag** [comment]
Concept ACK.

**Comment by fanquake** [comment]
Concept ACK
```
bash-3.2$ ../contrib/devtools/circular-dependencies.py {*,*/*,*/*/*}.{h,cpp}
Circular dependency: chain -> pow -> chain
Circular dependency: chainparamsbase -> util -> chainparamsbase
Circular dependency: checkpoints -> validation -> checkpoints
Circular dependency: init -> net_processing -> init
Circular dependency: init -> index/txindex -> init
Circular dependency: init -> rpc/server -> init
Circular dependency: init -> txdb -> init
Circular dependency: init -> validation -> init
Circular dependency: init -> validationinterface -> init
Circular dependency: random -> util -> random
Circular dependency: sync -> util -> sync
Circular dependency: txmempool -> policy/fees -> txmempool
Circular dependency: txmempool -> validation -> txmempool
Circular dependency: validation -> index/txindex -> validation
Circular dependency: validation -> validationinterface -> validation
Circular dependency: validation -> policy/policy -> validation
Circular dependency: qt/addresstablemodel -> qt/walletmodel -> qt/addresstablemodel
Circular dependency: qt/bantablemodel -> qt/clientmodel -> qt/bantablemodel
Circular dependency: qt/bitcoingui -> qt/utilitydialog -> qt/bitcoingui
Circular dependency: qt/bitcoingui -> qt/walletview -> qt/bitcoingui
Circular dependency: qt/bitcoingui -> qt/walletframe -> qt/bitcoingui
Circular dependency: qt/clientmodel -> qt/peertablemodel -> qt/clientmodel
Circular dependency: qt/paymentserver -> qt/walletmodel -> qt/paymentserver
Circular dependency: qt/recentrequeststablemodel -> qt/walletmodel -> qt/recentrequeststablemodel
Circular dependency: qt/sendcoinsdialog -> qt/walletmodel -> qt/sendcoinsdialog
Circular dependency: qt/transactiontablemodel -> qt/walletmodel -> qt/transactiontablemodel
Circular dependency: qt/walletmodel -> qt/walletmodeltransaction -> qt/walletmodel
Circular dependency: rpc/rawtransaction -> wallet/rpcwallet -> rpc/rawtransaction
Circular dependency: wallet/coincontrol -> wallet/wallet -> wallet/coincontrol
Circular dependency: wallet/fees -> wallet/wallet -> wallet/fees
Circular dependency: wallet/rpcwallet -> wallet/wallet -> wallet/rpcwallet
Circular dependency: wallet/wallet -> wallet/walletdb -> wallet/wallet
Circular dependency: txmempool -> validation -> policy/rbf -> txmempool
Circular dependency: txmempool -> validation -> validationinterface -> txmempool
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/addressbookpage
Circular dependency: qt/guiutil -> qt/walletmodel -> qt/optionsmodel -> qt/guiutil
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/signverifymessagedialog -> qt/addressbookpage
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/receivecoinsdialog -> qt/addressbookpage
Circular dependency: qt/guiutil -> qt/walletmodel -> qt/optionsmodel -> qt/intro -> qt/guiutil
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/sendcoinsdialog -> qt/sendcoinsentry -> qt/addressbookpage
```

**Comment by practicalswift** [comment]
Concept ACK

Very nice!

Suggested changes to make the output deterministic:

```
--- ../circular-dependencies.py	2018-05-14 08:50:58.691253505 +0200
+++ ../circular-dependencies-deterministic.py	2018-05-14 09:28:20.164794798 +0200
@@ -35,7 +35,7 @@

 # Iterate again, and build list of direct dependencies for each module
 # TODO: implement support for multiple include directories
-for arg in files.keys():
+for arg in sorted(files.keys()):
     module = files[arg]
     with open(arg, 'r') as f:
         for line in f:
@@ -49,7 +49,7 @@
 # Loop to find the shortest (remaining) circular dependency
 while True:
     shortest_cycle = None
-    for module in deps.keys():
+    for module in sorted(deps.keys()):
         # Build the transitive closure of dependencies of module
         closure = dict()
         for dep in deps[module]:
@@ -57,7 +57,7 @@
         while True:
             old_size = len(closure)
             old_closure_keys = list(closure.keys())
-            for src in old_closure_keys:
+            for src in sorted(old_closure_keys):
                 for dep in deps[src]:
                     if dep not in closure:
                         closure[dep] = closure[src] + [src]
```

**Comment by Empact** [comment]
nit: I would expect this to work from root, with a call like:
`contrib/devtools/circular-dependencies.py src/{*,*/*,*/*/*}.{h,cpp}` but it doesn't.

**Comment by sipa** [comment]
Yes, it can't. To deal with that it needs logic to actually resolve includes w.r.t. include directories; that's a TODO in the code. The README gives the correct invocation.

File: contrib/devtools/README.md [hunk 1]
```diff
@@ -194,3 +194,14 @@ It will do the following automatically:
...
+circular-dependencies.py
+========================
+
+Run this script from the root of the source tree (`src/`) to find circular dependencies in the source code.
+This looks only at which files include other files, treating the `.cpp` and `.h` file as one unit.
+
+Example usage:
+
+    cd .../src
+    ../contrib/devtools/circular-dependencies.py {*,*/*,*/*/*}.{h,cpp}
```
Comment (promag) [inline contrib/devtools/README.md]: nit `../contrib/devtools/circular-dependencies.py  **/*.{h,cpp}`.

**Review by promag — COMMENTED** [review]
Tested ACK 50c69b78.

👍 to @practicalswift suggestions.

Is there a use case to pass only a subset of the files? I mean, instead of passing the list of files, the script could traverse the entire source tree.

Suggestion, script could exit 0 if no circular dependencies are found.

File: contrib/devtools/README.md (hunk 1 above)
Comment (Empact) [reply contrib/devtools/README.md]: @promag FWIW on my system **/*.h does not capture immediate *.h files.

**Comment by Mickieki** [comment]
https://github.com/bitcoin/bitcoin/issues/13235

**Comment by ken2812221** [comment]
Tested ACK 50c69b78011c1bc55885ebfd216db60ed490ebea

**Comment by practicalswift** [comment]
ACK assuming output is made deterministic

Nit: Would be nice if the exit code indicated if circular dependencies were found

File: contrib/devtools/README.md (hunk 1 above)
Comment (promag) [reply contrib/devtools/README.md]: Ah 😝never mind then.

**Comment by sipa** [comment]
Made the ordering deterministic, and added an exit code.

**Comment by practicalswift** [comment]
utACK a7b295e91e4917495efe59948bae0ea554b7674c

**Comment by ken2812221** [comment]
utACK a7b295e

**Pull request merged**